### DISCUSSION TITLE
Move extensions into seperated folders

### DISCUSSION THREAD

**Discussion by Attacler** [directus/directus#5360] [Feature Requests]
Maybe it is an idea to make extensions happen in seperated directories.

Current approach:
extensions/hooks/Extension/index.js

Suggested approach:
extensions/Extension/hooks/index.js

With this approach you can basicly make extensions more "plugin" like, you can just copy 1 directory over and if you want to know what a extension needs/does, you can check it inside one directory instead of across all those differend directories where all hooks/layouts ect are present.

**Comment by danieljwestman** [comment]
From what I understand this is the upcoming approach concurrent with a npm extension routine 🚀

**Comment by rijkvanzanten** [comment]
@nickrum Thoughts on "reversing" that extension directory scoping?

**Comment by Kematia** [comment]
Hard agree on this one! Let's say we have a blog-alike "articles" module, it would supply it's own interfaces (ie: slugs), displays, migrations (setup articles_categories and articles_items) and hooks (create_slug). Having it scoped to articles/interface|hooks|displays|migrations would be much nicer!

**Reply by rijkvanzanten** [reply]
We'll have to take a good hard look at the structure within each extension then 🤔 Technically speaking, hooks/endpoints register multiple in a single file, while interfaces/displays/modules/etc are only 1 per JS file

**Reply by Attacler** [reply]
> We'll have to take a good hard look at the structure within each extension then  Technically speaking, hooks/endpoints register multiple in a single file, while interfaces/displays/modules/etc are only 1 per JS file

True but still, to me it seems like the most "extension"/pluggable system you could get when you just can pop in a directory and you have a extension istalled.
It probbaly will also be easier to setup and you could make extensions uploadable in zip format (a little like wordpress does)

**Reply by rijkvanzanten** [reply]
Yeah totally, though that doesn't really differ based on the "order" of the folders. The idea initially was that "an extension" was a single interface / hook etc that you could install how you describe it, that being said i agree that it makes more sense to have extensions be a single entity that can then render multiple _things_.

Just saying that we should figure out how to bundle/expose those, as it would also make more sense at that point to bundle all app extensions into a single entrypoint (to prevent loading overhead on the front-end), and (likewise) have the API ones combined, which should simplify the build step too 🤔  (cc @nickrum )

**Reply by Attacler** [reply]
> Yeah totally, though that doesn't really differ based on the "order" of the folders. The idea initially was that "an extension" was a single interface / hook etc that you could install how you describe it, that being said i agree that it makes more sense to have extensions be a single entity that can then render multiple _things_.
> 
> Just saying that we should figure out how to bundle/expose those, as it would also make more sense at that point to bundle all app extensions into a single entrypoint (to prevent loading overhead on the front-end), and (likewise) have the API ones combined, which should simplify the build step too  (cc @nickrum )

Maybe you can do it differently?
Basically keep the old way and add a folder called plugins or something that can have pluggable extensions?

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
